dash_labs/dashdown.py

```diff
@@ -0,0 +1,438 @@
+                    if not scope_creep:
+                        # make a copy
+                        code_scope = dict(**scope)
+                    else:
+                        # use the same dict that'll keep getting mutated
+                        code_scope = scope
+                    try:
+                        code = section.replace("```python", "").replace("```", "")
+                        app_card = _run_code(
+                            code,
+                            scope=code_scope,
+                            div_props=updated_props["app_div_props"],
+                        )
+                    except Exception as e:
+                        msg1 = f"\nError: {e} "
+                        msg2 = _get_first_lines(section, code_block, filename)
+                        print(msg1 + msg2)
+                        pass
+
+                # side by side on large screens
+                lg = 6 if updated_props["side_by_side"] else 12
+                if updated_props["code_first"]:
+                    code_display = [
+                        dbc.Col(code_card, width=12, lg=lg),
+                        dbc.Col(app_card, width=12, lg=lg),
+                    ]
+                else:
+                    code_display = [
+                        dbc.Col(app_card, width=12, lg=lg),
+                        dbc.Col(code_card, width=12, lg=lg),
+                    ]
+                reconstructed.append(dbc.Row(code_display))
+            else:
+                reconstructed.append(
+                    dbc.Row(dbc.Col(dcc.Markdown(section, **text_markdown_props)))
+                )
+        super().__init__(reconstructed)
+
+
+def _prohibited_props_check(props, prohibited_props, code_block=None, filename=None):
+    """
+    certain props may not be updated for security reasons and/or because they are updated by
+    the MarkdownAIO component
+    """
+
+    for prop in props:
+        if isinstance(props[prop], dict):
+            for p in props[prop]:
+                if p in prohibited_props[prop]:
+                    raise Exception(
+                        f" The `{p}` prop in `{prop}` may not be user-supplied in MarkdownAIO"
+                        + f" see code block# {code_block} in module {filename}"
+                        if code_block
+                        else ""
+                    )
+
+
+def _run_code(code, scope=None, div_props=None):
+    if scope is None:
+        scope = {}
+
+    # Replacements to get this working in an `exec` environment:
+    # 1) Remove language specifier included at beginning of markdown string
+
+    language = code[:6].lower()
+    if language.startswith("python"):
+        code = code[6:]
+    # 2) Remove the app instance in the code block otherwise app.callbacks don't work
+    code = _remove_app_instance(code)
+
+    if "app.layout" in code:
+        code = code.replace("app.layout", "layout")
+    if "layout" in code:
+        exec(code, scope)
```

Comment:
I'm feeling like we should lock this down a bit more and exclude folks from using exec within callbacks. 
for users, this means:
- `MarkdownAIO` can only be called outside of callbacks, meaning when the app starts up
- All of the `MarkdownAIO` files would need to be written in advance of the app starting up

so this wouldn't be allowed:
`pages/report.py`
```
dash.register_page(__name__)
def layout():
    return MarkdownAIO('report.md', exec_code=True)
```

But this would be allowed:
`pages/report.py`
```
dash.register_page(__name__)
layout = MarkdownAIO('report.md', exec_code=True)
```

I can't think of many examples where this would be insufficient and it would prevent users from seemingly innocent but very insecure code like:

`template_report.md`
```
# Report for {customer}

`` `python
dcc.Graph(figure=px.scatter(...))
`` `
```

```python
app.layout = html.Div([
    dcc.Dropdown(id='customer', options=['Acme', 'City']),
    html.Div(id='content')
])

@callback(Output('content', 'children'), Input('customer', 'value'))
def update(customer):
    with open('template_report.md', 'r') as f:
        content = f.read()
        
    customer_report = content.replace(customer=customer)
    with open('customer_report.md', 'w') as f:
        f.write(customer_report)
    return MarkdownAIO('customer_report.md', exec_code=True)
```

Reply:
The way to prevent exec from running in a callback would be to do:
```
def _exec(*args, **kwargs):
    # flask raises an error if flask.request.path is called outside of a request
    try:
        flask.request.path
    except RuntimeError as e:
        raise Exception('MarkdownAIO is being called with `exec_code=True` within a callback or flask request. This isn\'t supported. MarkdownAIO with exec can only be called when the app is starting')

    return exec(*args, **kwargs)
```

Reply:
When I tried this it block all `exec` s in the app.  Any other suggestions?